### ISSUE TITLE
[SOLVED] illegal hardware instruction error on Apple Silicon

### ISSUE THREAD

**Issue by alexjolig** [pozitronik/sinner#86]
I'm trying to run the project on my Mac Apple Silicon, I installed everything in a `venv` with python 3.10 and tried 

`venv/bin/python sin.py --source="path/foor.jpg" --target="path/john.jpg" --frame-processor=FaceSwapper
`
But I got this error:
`[1]    12026 illegal hardware instruction  venv/bin/python sin.py   --frame-processor=FaceSwapper`

Then I removed my `venv` and created another one with `python3.10-intel64` version which usually works for me in my Python projects in my Mac M2.

Then I tried to run the following:

`venv/bin/python3.10-intel64 sin.py --source="path/foor.jpg" --target="path/john.jpg" --frame-processor=FaceSwapper`

But again I got this error:

`[1]    12228 illegal hardware instruction  venv/bin/python3.10-intel64 sin.py   --frame-processor=FaceSwapper`

So what should I do to fix it?

**Comment by pozitronik** [comment]
>So what should I do to fix it?

Actually, I'm not sure. I don't have a Mac, nor have I seen an error like this before. But I can make an assumption that the cause might be related to ONNX Runtime on the M2 architecture.

So, I need you to run some diagnostics.

First, run
`venv/bin/python sin.py --source="path/foor.jpg" --target="path/john.jpg" --frame-processor=DummyProcessor`

This won't affect the images but will run all the processing code except for ONNX Runtime tasks.

Then run
`venv/bin/python sin.py --source="path/foor.jpg" --target="path/john.jpg" --frame-processor=FaceEnhancer`

This will run the face enhancing model using ONNX Runtime, showing us if the problem related to the model or the runtime.

Post the results here, and we'll try to figure out what to do next.

**Comment by alexjolig** [comment]
> Post the results here, and we'll try to figure out what to do next

Thanks for the reply. I tried ran this command:

`venv/bin/python sin.py --source="path/foor.jpg" --target="path/john.jpg" --frame-processor=DummyProcessor`

And I tried with different python runners, `python`, `python3.10`, `python3.10-intel64`. But I'm getting the same error:

`[1]    12657 illegal hardware instruction  venv/bin/python3.10-intel64 sin.py --source="path/foor.jpg"`

I also tried `FaceEnhancer` and got the same error again.

**Comment by pozitronik** [comment]
Run
`python -m pip list`
and post the output here. I need to check if the correct packages for your hardware are installed.

**Comment by alexjolig** [comment]
Package                      Version
---------------------------- ---------------
absl-py                      2.0.0
addict                       2.4.0
albumentations               1.3.1
astunparse                   1.6.3
basicsr                      1.4.2
cachetools                   5.3.2
certifi                      2023.7.22
charset-normalizer           3.3.1
colorama                     0.4.6
coloredlogs                  15.0.1
contourpy                    1.1.1
customtkinter                5.2.0
cycler                       0.12.1
Cython                       3.0.4
darkdetect                   0.8.0
easydict                     1.11
facexlib                     0.3.0
filelock                     3.13.0
filterpy                     1.4.5
flatbuffers                  23.5.26
fonttools                    4.43.1
future                       0.18.3
gast                         0.5.4
gfpgan                       1.3.8
google-auth                  2.23.3
google-auth-oauthlib         1.0.0
google-pasta                 0.2.0
grpcio                       1.59.0
h5py                         3.10.0
humanfriendly                10.0
idna                         3.4
imageio                      2.31.6
importlib-metadata           6.8.0
insightface                  0.7.3
Jinja2                       3.1.2
joblib                       1.3.2
keras                        2.14.0
kiwisolver                   1.4.5
lazy_loader                  0.3
libclang                     16.0.6
llvmlite                     0.41.1
lmdb                         1.4.1
Markdown                     3.5
MarkupSafe                   2.1.3
matplotlib                   3.8.0
ml-dtypes                    0.2.0
mpmath                       1.3.0
networkx                     3.2
numba                        0.58.1
numpy                        1.26.1
oauthlib                     3.2.2
onnx                         1.14.0
onnxruntime                  1.15.0
opencv-python                4.8.1.78
opencv-python-headless       4.8.1.78
opt-einsum                   3.3.0
packaging                    23.2
pathvalidate                 3.2.0
Pillow                       10.0.0
pip                          22.3.1
platformdirs                 3.11.0
prettytable                  3.9.0
protobuf                     4.24.4
psutil                       5.9.5
pyasn1                       0.5.0
pyasn1-modules               0.3.0
pyparsing                    3.1.1
python-dateutil              2.8.2
pyvirtualcam                 0.10.2
PyYAML                       6.0.1
qudida                       0.0.4
requests                     2.31.0
requests-oauthlib            1.3.1
rsa                          4.9
scikit-image                 0.22.0
scikit-learn                 1.3.2
scipy                        1.11.3
setuptools                   65.5.1
six                          1.16.0
sympy                        1.12
tb-nightly                   2.16.0a20231027
tensorboard                  2.14.1
tensorboard-data-server      0.7.2
tensorflow                   2.14.0
tensorflow-estimator         2.14.0
tensorflow-io-gcs-filesystem 0.34.0
termcolor                    2.3.0
threadpoolctl                3.2.0
tifffile                     2023.9.26
tk                           0.1.0
tomli                        2.0.1
torch                        2.0.1
torchvision                  0.15.2
tqdm                         4.66.1
typing_extensions            4.8.0
urllib3                      2.0.7
wcwidth                      0.2.8
Werkzeug                     3.0.1
wheel                        0.38.4
wrapt                        1.14.1
yapf                         0.40.2
zipp                         3.17.0

**Comment by pozitronik** [comment]
I don't see the `onnxruntime-silicon` package in the list, only `onnxruntime`. Did you use `pip install -r requirements-mac-arm64.txt` command to install required packages?

First, install the required package, using
```bash
python -m pip uninstall onnxruntime
python -m pip install onnxruntime-silicon
```
(do not forget to make sure your venv is active before installation).

Then, try to run the script, and tell if there's any differences.

I've also found some clues pointing this is might be related to the `tensorflow` support on the Apple Silicon architecture. You can check if it is by running
```bash
python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])))"
```
If you'll get a hardware error, you need to dig in that direction.

**Comment by alexjolig** [comment]
At first, I tried `pip install -r requirements-mac-arm64.txt` But it failed to install  `onnxruntime-silicon` because it said the package you try to install ia not found with the tag or something. I checked [here](https://pypi.org/project/onnxruntime/#files) and tried the other versions, no luck. That's why I chose to the approach with `python3.10-intel64` version.

anyways, I tested the `tensorflow` code you suggested, and I got the exact error for architecture. So I'm now sure it has something to do with it. 
I found some same issues in Stackoverflow about it, I'm test them out and be back with the results.

**Comment by alexjolig** [comment]
So following [this answer](https://stackoverflow.com/a/76000344/2263683) in stackoverflow solved my issue.

I'll also mention it here:

`python -m pip uninstall tensorflow-macos`
`python -m pip uninstall tensorflow-metal`

`SYSTEM_VERSION_COMPAT=0 pip install tensorflow-macos tensorflow-metal`


Thank you for your help.
